### PR TITLE
Gutenberg: Add editor styles to WordPress.com Simple sites' site editor

### PR DESCRIPTION
**Pull request by Copons**
Fixes https://github.com/Automattic/wp-calypso/issues/59501

#### Changes proposed in this Pull Request:

* Add a new method to inline styles in the block editor with the Core `add_editor_style` function.
* Add a new method to inline styles in the site editor.

#### The Issue

The site editor is currently rendered inside an `editor-canvas` iframe, which natively supports stylesheets added with `add_editor_style`.

To apply those enqueued with `wp_enqueue_style` to the iframed canvas, Gutenberg [copies](https://github.com/WordPress/gutenberg/blob/d257d326075fafeb473adcf16c7ed1d2b54ac384/packages/block-editor/src/components/iframe/index.js#L30-L92) the relevant stylesheets from the document's head and inlines them inside the iframe.

If the CSS is served from a different origin, this compatibility solution [fails](https://github.com/WordPress/gutenberg/blob/d257d326075fafeb473adcf16c7ed1d2b54ac384/packages/block-editor/src/components/iframe/index.js#L44-L46).

This is what is happening on Simple sites in production.

We have an aggressive optimization process that minifies, concatenates, and serves from a CDN all Jetpack assets.
The JP editor styles are enqueued in "the old way", and served from `s0.wp.com`, which breaks the Gutenberg compatibility approach.

The result is that JP blocks and editor extensions are missing all styles when rendered in the site editor.

#### The Fix (Kudos to @creativecoder)

`add_editor_style` single-handedly fixes the issue, and it also works with stylesheets provisioned by the CDN.
The function is not supposed to run with the enqueue hooks but works fine in `admin_init`.
It also automatically provides an RTL version of the stylesheet whenever needed.

We call `add_editor_style` through `Jetpack_Gutenberg::add_site_editor_style`, which handles most of the logic.

We'll need to keep the situation monitored, though, and come up with another solution before Core Gutenberg iframes the post editor too.

---

To get the full picture, please read this: paYJgx-1Kl-p2

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

There are no functional changes to Jetpack, so please follow the testing instructions in D74363-code.

### CODE REVIEW

**Comment by anomiex** [comment]
I'm finding myself a bit confused here. [add_editor_style](https://developer.wordpress.org/reference/functions/add_editor_style/) is documented as being for _themes_ to add styles to the _TinyMCE_ editor, not the block editor?

As far as I can tell from looking at [WordPress Core's block editor API documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/), it seems they want the style handle to be declared as being for the block editor via `register_block_type()` (specifically a `style` or `editor_style` in `$args`). Which also seems like it may be wanting to be done per block rather than having a giant "editor" blob like Jetpack currently does.

> We'll need to keep the situation monitored, though, and come up with another solution before Core Gutenberg iframes the post editor too.

Overall this smells like a hack rather than an actual fix. A hack is ok, but maybe we should put the hack directly in the `Jetpack_Gutenberg::add_site_editor_style` method being added here and document it as such rather than putting it in Assets like it's something people should actually be using long term?

A non-hack Assets function seems like it would be building the args for `register_block_type()` (including the scripts too, not just styles) and calling that function.

**Comment by Copons** [comment]
@anomiex `add_editor_style` is indeed built with themes in mind: it looks for the stylesheet in the theme folder.
Although, it also works with absolute URLs — as we do in this particular workaround.

You are also correct when you say this is a hack, and it should be temporary until we find a more appropriate solution! 😄 
The problem with registering block assets as suggested by Gutenberg is that _now_ JP builds all of them together with general-purpose editor extensions (such as the upgrade nudge banner, which was the issue that originated this change and uncovered the missing block styles).

The Metadata API (and many, many other block improvements) has been added in 5.8, which means it has only just recently become available to JP, while JP blocks have been typically developed in 2020 with some updates in 2021.

This kind of change is outside of my knowledge area and is definitely outside of the scope of a high priority Dotcom bug fix. 🙂 
The workaround proposed here is effective without introducing regressions (AFAICS, but never say never 😄).
With this in mind, I agree that updating the Assets package is probably unwarranted.
I'll update the PR consolidating the code into `Jetpack_Gutenberg`, clarifying the purpose of the fix.

Does Jetpack have a naming convention for functions that are not recommended for general use (something like `unstable__`)?

**Comment by creativecoder** [comment]
Actually, I don't think this is a hack. It seems that Gutenberg is reviving the `add_editor_style` function for declaring block styles that should load in the editor.

Here you can see the Site editor getting the editor styles (via the `get_block_editor_theme_styles` function, which uses the `$editor_styles` global, https://github.com/WordPress/wordpress-develop/blob/1712bfcf79097df6766f093f3695c84f058e2666/src/wp-admin/site-editor.php#L56)

And there are comments on Gutenberg issues recommending using `add_editor_style` for adding styles to the Site editor iframe, both for themes and plugins: https://github.com/WordPress/gutenberg/issues/33212#issuecomment-879447803

I believe the plan is to eventually iframe the Post editor as well, so it seems that loading block styles this way will only become more necessary.

**Review by creativecoder — COMMENTED** [review]
I wouldn't tie the naming to the Site editor. I think any of the editors may end up using an iframe (post, site, navigation, widgets), so I would keep this flexible.

For instance, the post editor loads these styles as well (https://github.com/WordPress/wordpress-develop/blob/1712bfcf79097df6766f093f3695c84f058e2666/src/wp-admin/edit-form-blocks.php#L194) and may lean on theme for inlining styles in an iframe. Anywhere blocks are used in a block editor, we may need to load these.

File: projects/plugins/jetpack/class.jetpack-gutenberg.php [hunk 1]
```diff
@@ -699,6 +699,54 @@ public static function enqueue_block_editor_assets() {
...
 
+	/**
+	 * Add the Gutenberg editor stylesheet to the site editor.
+	 * This workaround is only supposed to be used by WordPress.com Simple sites.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return void
+	 */
+	public static function add_site_editor_style() {
```
[the remark this answers is not in the document]
Comment (Copons) [reply projects/plugins/jetpack/class.jetpack-gutenberg.php:710]: It's only hooked in D74363-code.
Unless there's a specific need for doing it on Jetpack too (AFAIK it should work fine as it is right now), I'd avoid it. 🤔

**Comment by Copons** [comment]
> Actually, I don't think this is a hack.

@creativecoder Apologies, I could have detailed more.

IMHO the appropriate solution would be to build and enqueue block assets on a per-block basis through the block's metadata, as suggested above.
It would require rewriting the whole JP's build logic (it could be trivial? I don't know, but I'd rather not add the burden of regressions on me 😄 ) to split the files.
It would also leave the editor extensions assets without a good place — and yes, AFAICS at the moment the only way to enqueue non-block styles for iframed editors is `add_editor_style`.

I mean, we could call this hack or workaround, but either way, I wouldn't consider this a permanent solution, instead just a stopgap to take time to address the whole picture without fires. 🙂

**Comment by anomiex** [comment]
> Actually, I don't think this is a hack. It seems that Gutenberg is reviving the `add_editor_style` function for declaring block styles that should load in the editor.

What is Gutenberg intending to do about things like this code disabling some other styles when `$editor_styles` is set? https://github.com/WordPress/WordPress/blob/e8e20a282371286e958cca39ced6c9b8a76641fc/wp-includes/script-loader.php#L1550-L1553
Or the check here for theme support before processing `$editor_styles`? https://github.com/WordPress/WordPress/blob/e8e20a282371286e958cca39ced6c9b8a76641fc/wp-includes/block-editor.php#L522

> IMHO the appropriate solution would be to build and enqueue block assets on a per-block basis through the block's metadata, as suggested above.
> It would require rewriting the whole JP's build logic (it could be trivial? I don't know, but I'd rather not add the burden of regressions on me smile ) to split the files.

I can help with figuring out the build logic in Jetpack. I don't know much about the actual blocks though.

**Comment by Copons** [comment]
@anomiex I don't know Core plans, but at least in regards to this:

> Or the check here for theme support before processing $editor_styles?

`add_editor_style` automatically adds support for `editor-styles` when used:

https://github.com/WordPress/wordpress-develop/blob/871afd0205214a745d85d20349a1f2bc0fe11895/src/wp-includes/theme.php#L2086

**Comment by anomiex** [comment]
> `add_editor_style` automatically adds support for `editor-styles` when used:

Ah, good to know. But then does that have any side effects elsewhere? I don't see any at a quick grep, but it is apparently output in the REST API.

File: projects/plugins/jetpack/class.jetpack-gutenberg.php [hunk 2]
```diff
@@ -699,6 +699,56 @@ public static function enqueue_block_editor_assets() {
...
+	 *
+	 * @return void
+	 */
+	public static function add_iframed_editor_style() {
+		if ( ! self::should_load() ) {
+			return;
+		}
+
+		global $pagenow;
+		$allowed_pages       = array( 'admin.php', 'themes.php' );
```
Comment (creativecoder) [inline projects/plugins/jetpack/class.jetpack-gutenberg.php:718]: Note that the site editor url can also be /wp-admin/site-editor.php in WP 5.9 without Gutenberg active, so we should probably account for that as well.

File: projects/plugins/jetpack/class.jetpack-gutenberg.php [hunk 3]
```diff
@@ -699,6 +699,56 @@ public static function enqueue_block_editor_assets() {
 		);
 	}
 
+	/**
+	 * Add the Gutenberg editor stylesheet to iframed editors, such as the site editor,
+	 * which don't have access to stylesheets added with `wp_enqueue_style`.
+	 *
+	 * This workaround is only supposed to be used by WordPress.com Simple sites.
```
Comment (creativecoder) [inline projects/plugins/jetpack/class.jetpack-gutenberg.php:706]: It's quite possible this is needed when other plugins are used that do asset optimization in specific ways, so I wouldn't say it's only for Simple Sites.

**Review by creativecoder — COMMENTED** [review]
✅ for this fixing the WP.com issue with Jetpack block styles in the Site editor (when combined with D74363-code). I left comments on both with a couple of tweaks, then let's get this fix out there as soon as we can!

File: projects/plugins/jetpack/class.jetpack-gutenberg.php (hunk 2 above)
Comment (Copons) [reply projects/plugins/jetpack/class.jetpack-gutenberg.php:718]: TIL!

**Review by sdixon194 — APPROVED** [review]

**Comment by Copons** [comment]
r239822-wpcom

**Pull request merged**